openstf/pipeline/train_predict_backtest.py

```diff
@@ -0,0 +1,135 @@
+    trained_models_folder: Union[str, Path],
+    save_figures_folder: Union[str, Path],
+    training_horizons: List[float] = None,
+) -> pd.DataFrame:
+    """Pipeline for a back test.
+
+        DO NOT USE THIS PIPELINE FOR OPERATIONAL FORECASTS
+
+    Args:
+        pj (dict): Prediction job.
+        input_data (pd.DataFrame): Input data
+        trained_models_folder:
+        trained_models_folder (Path): Path where trained models are stored
+        save_figures_folder (Path): path were reports (mostly figures) about the training procces are stored
+        training_horizons (list): Horizons to train on in hours.
+            These horizons are also used to make predictions (one for every horizon)
+
+    Returns:
+        forecast (pandas.DataFrame)
+
+    """
+    if training_horizons is None:
+        training_horizons = DEFAULT_TRAIN_HORIZONS
+
+    # Validate and clean data
+    validated_data = validation.clean(validation.validate(input_data))
+
+    # Check if sufficient data is left after cleaning
+    if not validation.is_data_sufficient(validated_data):
+        raise ValueError(
+            f"Input data is insufficient for {pj['name']} "
+            f"after validation and cleaning"
+        )
+
+    # Add features
+    data_with_features = TrainFeatureApplicator(
+        horizons=training_horizons, feature_names=pj["feature_names"]
+    ).add_features(validated_data)
+
+    # Split data
+    train_data, validation_data, test_data = split_data_train_validation_test(
+        data_with_features.sort_index(axis=0), test_fraction=0.15, back_test=True
+    )
+
+    # Create relevant model
+    model = ModelCreator.create_model(pj)
+
+    # split x and y data
+    train_x, train_y = train_data.iloc[:, 1:-1], train_data.iloc[:, 0]
+    validation_x, validation_y = (
+        validation_data.iloc[:, 1:-1],
+        validation_data.iloc[:, 0],
+    )
+
+    # Configure evals for early stopping
+    eval_set = [(train_x, train_y), (validation_x, validation_y)]
+
+    model.set_params(**pj["hyper_params"])
+    model.fit(
+        train_x,
+        train_y,
+        eval_set=eval_set,
+        early_stopping_rounds=DEFAULT_EARLY_STOPPING_ROUNDS,
+        verbose=False,
+    )
+    logging.info("Fitted a new model, not yet stored")
+
+    # Do confidence interval determination
+    model = StandardDeviationGenerator(
+        pj, validation_data
+    ).generate_standard_deviation_data(model)
+
+    # Report about the training procces
+    report = Reporter(pj, train_data, validation_data, test_data).generate_report(model)
+
+    # Save model
+    serializer = PersistentStorageSerializer(trained_models_folder)
+    serializer.save_model(model, pid=pj["id"])
+
+    # Save reports/figures
+    report.save_figures(save_path=save_figures_folder)
+
+    # Load most recent model for the given pid
+    model = PersistentStorageSerializer(
+        trained_models_folder=trained_models_folder
+    ).load_model(pid=pj["id"])
```

Comment:
Do you think it's needed to save a model and reports when performing a backtest?
Possibly we can do without storing a model and subsequently loading it back in when making the backtest.
Perhaps storing the model and reports is nice feature, but then I would not use the default locations, but for example to require this to be specified as an argument in the function.
What do you think?

Reply:
I was thinking a bit of the same. It seems a bit inflexible to always require to store stuff to disk by default. Could/should we not go for a bit more flexibility and less redundancy/duplication by having a structure like this:

* backtest pipeline
  * `train_model_pipeline()` (reused/calls `train_model.train_model_pipeline_core`)
  * `create_forecast_pipeline()` (reused/calls `create_forecast.create_forecast_pipeline_core`)
  * `train_forecast_pipeline()` (calls the above pipelines)

Reply:
True, I removed the saving of the model and the reporting in the pipeline. I made the pipeline also return the model such that the user can save it him/herself.